sdk/identity/azure-identity/azure/identity/_credentials/certificate.py

```diff
@@ -48,6 +48,11 @@ def get_token(self, *scopes, **kwargs):  # pylint:disable=unused-argument
         token = self._client.get_cached_access_token(scopes, query={"client_id": self._client_id})
         if not token:
             token = self._client.obtain_token_by_client_certificate(scopes, self._certificate, **kwargs)
+        elif self._client.should_refresh(token):
+            try:
+                self._client.obtain_token_by_client_certificate(scopes, self._certificate, **kwargs)
+            except Exception:  # pylint: disable=broad-except
+                pass
```

Comment:
This logic:

~~~ python
if not token:
  # get new token
elif should_refrsesh:
  try:
     # get new token
  except Exception:
     # swallow
~~~

seems to be present in most if not all the credentials. Perhaps it could be moved into a base or mixin, and have the implementation just provide a callback or an override for the `# get new token` functionality?

Reply:
Agreed. But different credentials have different ways to refresh/redeem tokens. So I have not found a clean way to do it. (We don't have a shared base class)

Reply:
What do you think of something like this:
```py

class CredentialBase(ABC):
    def __init__(self, **kwargs):
        self._client = AadClient(...)

    def _get_token_impl(*scopes, **kwargs):
        if not scopes:
            raise ValueError('"get_token" requires at least one scope')

        token = self._client.get_cached_access_token(scopes)
        if not token:
            token = self._request_token(scopes, **kwargs)
        elif self._client.should_refresh(token):
            try:
                self._request_token(scopes, **kwargs)
            except Exception:  # pylint:disable=broad-except
                pass
        return token

    @abc.abstractmethod
    def _request_token(self, *scopes, **kwargs):
        pass

class Credential(CredentialBase):
    def get_token(*scopes, **kwargs):
        """relevant user-facing docstring"""
        return self._get_token_impl(*scopes, **kwargs)

    def _request_token(*scopes, **kwargs):
        """get a new token according to this credential's personal idiom"""
        ...
```

Reply:
Do you mean we make a shared credential base?

I would like to have it into a separate issue/PR as code refactoring.

Reply:
I think when we do refactoring by adding a shared class for all credentials, we can do further than only this. But I don't want to rush it right before a release.

Reply:
#12601